[Shell-Parse] Use \STX and \ETX for quoting

lierdakil opened the pull request:
Also drops escaping characters.

lierdakil committed:
[Shell-Parse] Use \STX and \ETX for quoting

Also drops escaping

DanielG commented:
I dunno we're just shifting the problem from one set of characters that don't work to another. I'd really rather unambiguously prefix the commands with something to activate the escaping in a way that doesn't break anything.

lierdakil commented:
At the very least, I can say that `\STX` and `\ETX` won't appear in Haskell code. Probably not in filenames either (stranger things happen, but at least ghc won't accept relative paths with ASCII control characters as module path). Besides, those make sense semantically, and working with them from code is much easier than with double quotes or other characters. It's possible to input those in POSIX shell with `^V^B` and `^V^C` as well. So from my point of view, it's rather tempting.

If you insist, I can add a toggle. Something along the lines of this:
If first character in command line is `\ESC`, it is ignored, but `\STX` and `\ETX` are treated as quotes. Otherwise, they are treated as literal characters.
If you _really_ want double quotes, we could add those as well, using different toggle. But interfacing ghc-mod from code, it doesn't really matter what characters we use, does it?

DanielG commented:
Oh I'm perfectly happy with using STX and ETX as the escape characters, I'd just like some way to enable their use optionally. Maybe even just a prefix command that passes say a different argument parser function to the real command parser would do. Before I was thinking of adding some flag that's easy to identify unambiguously from the old command line format but I've thought about it a bit and just adding a new command that can be used as a prefix to any other command would probably be more compatible with the old behavior.

So for eample to enable escaping the fronted would send `escape check \STXFoo Bar/Main.hs\ETX` instead of just `check`. It's easy to check for the "escape" command in the parser code since it just involves another branch in the command "table" that calls itself recursively while passing a different argument parser. A simple implementation would probably not detect repeated uses of `escape` so `escape escape foo ..` would still be valid but I think that's OK.

lierdakil commented:
I see no real sense in using literal "escape" word, since we can't use optparse-applicative for it anyways. A prefix character (e.g. backslash, or `\ESC` as previously suggested) is easier to check and implement.

DanielG commented:
I'm not sure I understand. Why can't we use optparse for it?

I'm not suggesting we escape everything with the word "escape" but just to have it as an indicator whether the old or new argument splitting behavior should be used so we can be super sure we don't break anything and have the ability to change it yet again if that turns out to be needed.

lierdakil commented:
Okay, a little reminder. Optparse-applicative works with a list of arguments, as given by `getArgs`. In this case, escaping is handled by shell.

When we readLine from stdin, as with interactive mode, we don't have a list of strings, we only have one string. So we can't pass it to optparse-applicative yet -- we need to separate it into individual tokens. And token separation behavior would depend on prefix.

So, we can't use optparse-applicative to control escaping behavior. Only thing we can possibly do is pattern-match on start of string (or guard with `stripPrefix`). One character makes more sense in this context.

lierdakil committed:
[Shell-escape] Escape toggle with \ prefix

E.g.
    check file.hs
will treat quote characters as literal characters, while
    \check file.hs
will assume quoting behavior

Backslash will be dropped, naturally.

DanielG commented:
Ah right, sorry. Been playing around with real monadic parsers too much recently ;) But in any case doing the prefix match should also be easy.

Why just one character though? If we control the beginning of the string it doesn't really matter how much we put there. Worst case we can always just do something like `all (uncurry (==)) $ zip line "escape "`. I'd like this to not be too obscure, our wire protocol is already bad enough let's not make it even more weird.

Either way I just want a way to swap this out for something else if it turns out to be a problem (_cough_ _cough_ windos).

lierdakil commented:
Well, backslash toggle doesn't make this much more obscure than a word toggle IMO.

lierdakil commented:
See last commit.

DanielG commented:
Not for now, no. But what if we have to change the escaping method. I'd call the one you're proposing `"ascii-escape"` and if we need or want to change it we can call the next one something like `"quote-escape"` (if it uses quotes again for some reason). It just makes it more obvious what that is for.

Imagine if you've never seen ghc-mod's on wire protocol and you see a few commands and they look more or less like shell commands so you assume that's how the parser works but then you see `"\check ...` and you're like WTF? That's what I mean by "obscure" maybe inconsistent would be a better word. We're using something the user/developer already knows namely shell-ish commands so it breaks convention to have something that's entirely unlike that. (though the STX/ETX stuff is already in that category at least it's more or less obvious from the "ascii-escape" prefix something special is happening).

lierdakil committed:
[Shell-escape] 'ascii-escape ' prefix toggle

lierdakil commented:
Okay.

lierdakil committed:
[Shell-escape] Add empty input string test

DanielG commented:
Merged. (I made a new branch for the upcoming release so I don't have to keep rebasing the other stuff that's not going in there btw)
